Add Formless component template file

JordanMartinez opened the pull request:
## What does this pull request do?

Fixes #49

Also changes `halogen-storybook`'s branch from `halogen5` to `master` as the code wouldn't compile otherwise.

## Where should the reviewer start?

Read the 'other notes' section, skim the code and then answer my questions below

## How should this be manually tested?

AFAIK, we should only verify that this code compiles in the CI. 

## Other Notes:

There's a number of things that aren't correct in this PR with regard to the data type, `Name`.
- the output field of the row type isn't `Name` as it should be but `String`
- there might be a few type classes that `Name` unnecessarily implements (can't remember)
- ~there isn't an optic for `FieldError`, so the usage of `_Error` will likely stop compilation~
- there's no example of the `Initial` type class and how one would could implement it for `Name`

The halogen component part is basically the uncommented version of the template file. Since it's also lacking in a few other ways (more modular pieces that show how to use some of its API?), it would need to be completed before we could document it better. I'm not sure what should be included here and what should be left out because the `examples` folder already exists.

Questions I have:
- What else is `st.form` used for?
- I think the commented version should include 'additional state' as an example of its usage whereas the uncommented one should not. Your thoughts?
- I think this file should be separately licensed to ensure that one who uses it (since that's what it's for, right?) don't have to worry about satisfying such license constraints. Your thoughts?

Lastly, there's documenting each part in a clear manner.

JordanMartinez committed:
Added Formless component template

JordanMartinez committed:
Further cleanup template

JordanMartinez commented:
@thomashoneyman Mind taking a look at this sometime soon so I can finish it up?

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+import Template.DataAndValidation (NAME_FIELD, _name, minLength)
+import Type.Row (type (+))
+
+type FormFieldsRow f =
```
I quite like this!

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+      ]
+
+  handleEvent :: F.Event Form AddedState
+              -> F.HalogenM Form AddedState Action ChildSlots Message MonadType Unit
```
Mind dropping these down for indentation?

```purs
  handleEvent 
    :: F.Event Form AddedState
    -> F.HalogenM Form AddedState Action ChildSlots Message MonadType Unit
```

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+
+      -- We won't do this here, but this is how most will handle a form
+      -- submission: raise it as an event to their parent.
+      --      H.raise formFiels
```
formFields*

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+type Message = FormFields
+type ChildSlots =
+  ()
+type MonadType = Aff
```
I think I'd rather this remain polymorphic, `m`, as that's the most common way the component is used in practice. Then just `component :: forall m. MonadAff m => ... m`

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+  , initialize = Just Initialize
+  , finalize = Just Finalize
+  , receive = Just <<< Receive
+  }
```
I think each of these (`mkInput` included) could use a one-line description of what it's for, just as a reminder. If you would like, I can add those in for you, or feel free to do so. 

Something simple like:

```purs
-- Decide what, if anything, to do when Formless events occur. If you would like to raise 
-- events as messages, then use `F.raiseResult` as your `handleEvent` function.
handleEvent :: ...
```

thomashoneyman commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,129 @@
+        [ HP.type_ InputText
+        , HP.placeholder "Michael"
+        , HP.value (F.getInput _name st.form) -- access one value in form
+        , HE.onValueInput (Just <<< F.setValidate _name)
```
Perhaps a comment here, too:

```suggestion
        , HE.onValueInput (Just <<< F.setValidate _name) -- set the value of _name, then validate it
```

thomashoneyman commented:
Thanks for adding this @JordanMartinez -- I'm happy to merge with a couple of small tweaks I mentioned above. This is quite nice. I'll debug the CI failure, but it's not to do with your PR and so I am happy to merge even if I haven't quite fixed that yet.

JordanMartinez commented:
I hope to get to this sometime tonight.

JordanMartinez committed:
Fix indentation for template

JordanMartinez committed:
Fix typo in template example

JordanMartinez committed:
Template file: convert MonadType to m with MonadAff constraint

JordanMartinez committed:
Template file: document handleEvent function

JordanMartinez committed:
Temlate file: document mkInput

JordanMartinez committed:
Template file: more clearly document getInput and setValidate functions

JordanMartinez commented:
@thomashoneyman I've addressed all your comments except for documenting more parts. I think that might be better handled by you since you're more familiar with this library than I am.

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+    -> F.ComponentHTML Form Action ChildSlots m
+  render st =
+    HH.div_
+      [ HH.p_ [ HH.text $ "Validity: " <> show st.validity ]
```
Indicates whether the form's values are valid (i.e. validation has passed for all fields)

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+  render st =
+    HH.div_
+      [ HH.p_ [ HH.text $ "Validity: " <> show st.validity ]
+      , HH.p_ [ HH.text $ "Dirty: " <> show st.dirty ]
```
Indicates whether the form is currently being modified/validated.

thomashoneyman replied:
Indicates whether any field in the form has been changed from its initial state

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+    HH.div_
+      [ HH.p_ [ HH.text $ "Validity: " <> show st.validity ]
+      , HH.p_ [ HH.text $ "Dirty: " <> show st.dirty ]
+      , HH.p_ [ HH.text $ "Being Submitted: " <> show st.submitting ]
```
Indicates whether the 'submit' button has been clicked and the form's content is still being validated one last time before submission is accepted.

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+      [ HH.p_ [ HH.text $ "Validity: " <> show st.validity ]
+      , HH.p_ [ HH.text $ "Dirty: " <> show st.dirty ]
+      , HH.p_ [ HH.text $ "Being Submitted: " <> show st.submitting ]
+      , HH.p_ [ HH.text $ "Number of Errors: " <> show st.errors ]
```
Indicates the number of errors due to validation failing that need to be fixed before submission is accepted

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+      , HH.p_ [ HH.text $ "Dirty: " <> show st.dirty ]
+      , HH.p_ [ HH.text $ "Being Submitted: " <> show st.submitting ]
+      , HH.p_ [ HH.text $ "Number of Errors: " <> show st.errors ]
+      , HH.p_ [ HH.text $ "Number of Submit attempts: " <> show st.submitAttempts ]
```
Indicates the number of times user has attempted to submit the form

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+      , HH.p_ [ HH.text $ "Being Submitted: " <> show st.submitting ]
+      , HH.p_ [ HH.text $ "Number of Errors: " <> show st.errors ]
+      , HH.p_ [ HH.text $ "Number of Submit attempts: " <> show st.submitAttempts ]
+      , HH.p_ [ HH.text $ "Additional state was: " <> show st.additionalState ]
```
This is how we refer to the additional labels we added to the form's state in `AddedState`.

thomashoneyman replied:
We can also refer to any additional labels we used to extend the form's state; in this case, that means any field from our `AddedState` type.

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+    :: F.Event Form AddedState
+    -> F.HalogenM Form AddedState Action ChildSlots Message m Unit
+  handleEvent = case _ of
+    F.Submitted formContent -> do
```
Indicates that the form has been successfully submitted.

JordanMartinez commented on template/FormlessTemplate.purs:
```diff
@@ -0,0 +1,148 @@
+
+      -- This line exists so the code compiles.
+      pure unit
+    F.Changed formState -> do
```
Indicates that the form's content has been changed. This event is triggered anytime a field is changed, whether it passes validation or not.

JordanMartinez reviewed (commented):
I'm not sure what you had in mind for adding comments to each part, but this is what I would put. Feel free to correct/revise, etc. and I'll update it to what we agree upon here and push the changes following that.

thomashoneyman commented:
I've added a thumbs-up to your suggestions which I think are good as-is, and I've added my own version for ones that I'd tweak. On the whole you're almost spot-on with what I'd write. Thanks!

thomashoneyman commented:
CI has been fixed thanks to @linearray in #54, so once your updates here are applied I'll merge this in. Thanks again!

JordanMartinez committed:
Use comments to further document template file

JordanMartinez commented:
Done! Just waiting on CI now.

thomashoneyman commented:
@JordanMartinez mind rebasing or merging the current `halogen5` branch into yours to get the CI passing? Thanks!

thomashoneyman commented:
Actually -- don't worry about it, I'll just merge and let things run there. Thanks again!
